### PR TITLE
spark with new labeledpoint

### PR DESCRIPTION
**Pull request by CodingCat**

### CODE REVIEW

**Comment by CodingCat** [comment]
@tqchen , please help to review and check if I use Rabit correctly

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -17,41 +17,64 @@
...
+      }.cache()
+    } else {
+      null
+    }
+  }
+
+  def train(config: Config, trainingData: RDD[LabeledPoint], obj: ObjectiveTrait = null,
+      eval: EvalTrait = null): Option[XGBoostModel] = {
+    import DataUtils._
     val numWorkers = config.getInt("numWorkers")
```
Comment (tqchen) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: This should not come from config, instead use trainigData.numPartitions?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 2]
```diff
@@ -17,41 +17,64 @@
...
+    val sc = trainingData.sparkContext
+    val tracker = new RabitTracker(numWorkers)
+    if (tracker.start()) {
+      // TODO: build configuration map from config
+      val xgBoostConfigMap = new HashMap[String, AnyRef]()
+      val boosters = buildDistributedBoosters(trainingData, xgBoostConfigMap, numWorkers, round,
+        obj, eval)
+      // force the job
+      sc.runJob(boosters, (boosters: Iterator[Booster]) => boosters)
+      tracker.waitFor()
```
Comment (tqchen) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: do not need to wait here, the spark job will block until job finishes

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 3]
```diff
@@ -17,41 +17,64 @@
...
-    boosters.first()
+    val sc = trainingData.sparkContext
+    val tracker = new RabitTracker(numWorkers)
+    if (tracker.start()) {
+      // TODO: build configuration map from config
+      val xgBoostConfigMap = new HashMap[String, AnyRef]()
+      val boosters = buildDistributedBoosters(trainingData, xgBoostConfigMap, numWorkers, round,
+        obj, eval)
+      // force the job
+      sc.runJob(boosters, (boosters: Iterator[Booster]) => boosters)
```
Comment (tqchen) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Maybe a pesudo reduction is better, so the operations is still lazy

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 4]
```diff
@@ -17,41 +17,64 @@
...
-    }.cache()
-    // force the job
-    sc.runJob(boosters, (boosters: Iterator[Booster]) => boosters)
-    // TODO: how to choose best model
-    boosters.first()
+    val sc = trainingData.sparkContext
+    val tracker = new RabitTracker(numWorkers)
+    if (tracker.start()) {
+      // TODO: build configuration map from config
+      val xgBoostConfigMap = new HashMap[String, AnyRef]()
```
Comment (tqchen) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Let us use the same interface as in XGBoostScala, Map[String, AnyRef]

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 5]
```diff
@@ -17,41 +17,64 @@
...
-        val dMatrix = new DMatrix(new JDMatrix(dataBatches, null))
-        Iterator(SXGBoost.train(xgBoostConfigMap, dMatrix, round, watches = null, obj, eval))
-    }.cache()
-    // force the job
-    sc.runJob(boosters, (boosters: Iterator[Booster]) => boosters)
-    // TODO: how to choose best model
-    boosters.first()
+    val sc = trainingData.sparkContext
+    val tracker = new RabitTracker(numWorkers)
+    if (tracker.start()) {
```
Comment (tqchen) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Assert tracker spart

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 6]
```diff
@@ -17,41 +17,64 @@
...
+      trainingData.repartition(numWorkers).mapPartitions {
+        trainingSamples =>
+          Rabit.init(new java.util.HashMap[String, String]() {
+            put("DMLC_TASK_ID", TaskContext.getPartitionId().toString)
+          })
+          val dMatrix = new DMatrix(new JDMatrix(trainingSamples, null))
+          val booster = SXGBoost.train(xgBoostConfMap, dMatrix, round,
+            watches = new HashMap[String, DMatrix], obj, eval)
```
Comment (tqchen) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: we can watch training, but not test

**Pull request merged**